File: lib/models/document.dart
class CourseDocument {
  final String id;
  final String courseId;
  final String uploaderId;
  final String uploaderName;
  final String title;
  final String description;
  final String fileType;
  final String fileSize;
  final DateTime uploadDate;
  final bool isPreviousSemester;
  final String semester;
  final int downloads;

  const CourseDocument({
    required this.id,
    required this.courseId,
    required this.uploaderId,
    required this.uploaderName,
    required this.title,
    required this.description,
    required this.fileType,
    required this.fileSize,
    required this.uploadDate,
    this.isPreviousSemester = false,
    this.semester = 'Spring 2026',
    this.downloads = 0,
  });

  factory CourseDocument.fromJson(Map<String, dynamic> json) {
    return CourseDocument(
      id: json['id'] ?? json['_id'] ?? '',
      courseId: json['courseId'] ?? '',
      uploaderId: json['uploaderId'] ?? '',
      uploaderName: json['uploaderName'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      fileType: json['fileType'] ?? '',
      fileSize: json['fileSize'] ?? '',
      uploadDate: DateTime.tryParse(json['uploadDate'] ?? '') ?? DateTime.now(),
      isPreviousSemester: json['isPreviousSemester'] ?? false,
      semester: json['semester'] ?? 'Spring 2026',
      downloads: json['downloads'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'courseId': courseId,
    'uploaderId': uploaderId,
    'uploaderName': uploaderName,
    'title': title,
    'description': description,
    'fileType': fileType,
    'fileSize': fileSize,
    'uploadDate': uploadDate.toIso8601String(),
    'isPreviousSemester': isPreviousSemester,
    'semester': semester,
    'downloads': downloads,
  };
}

File: backend/models/Document.js
const mongoose = require('mongoose');

const documentSchema = new mongoose.Schema({
  courseId:      { type: String, required: true },
  uploaderId:   { type: String, required: true },
  uploaderName: { type: String, required: true },
  title:        { type: String, required: true },
  description:  { type: String, default: '' },
  fileType:     { type: String, default: 'PDF' },
  fileSize:     { type: String, default: '0 KB' },
  mimeType:     { type: String, default: 'application/octet-stream' },
  fileData:     { type: String, default: null }, // base64 encoded file content
  extractedText: { type: String, default: '' }, // text extracted from PDFs for chatbot RAG
  isPreviousSemester: { type: Boolean, default: false },
  semester:     { type: String, default: 'Spring 2026' },
  downloads:    { type: Number, default: 0 },
}, { timestamps: true });

documentSchema.methods.toJSON = function() {
  const obj = this.toObject();
  obj.id = obj._id.toString();
  obj.uploadDate = obj.createdAt;
  delete obj._id;
  delete obj.__v;
  return obj;
};

documentSchema.index({ courseId: 1 });

module.exports = mongoose.model('Document', documentSchema);
